components/core/src/clp/NetworkReader.cpp

```diff
@@ -0,0 +1,351 @@
+    }
+    m_downloader_thread = std::make_unique<DownloaderThread>(*this, offset, disable_caching);
+    m_downloader_thread->start();
+}
+
+NetworkReader::~NetworkReader() {
+    if (is_download_in_progress()) {
+        // Abort the download so the downloader thread can destroy the CURL resources and exit.
+        submit_abort_download_request();
+    }
+
+    while (is_downloader_thread_running()) {
+        // We could use sleep instead of busy waiting
+    }
+}
+
+auto NetworkReader::try_get_pos(size_t& pos) -> ErrorCode {
+    if (0 != m_offset) {
+        if (State::InProgress != get_state_code()) {
+            auto const curl_return_code{get_curl_return_code()};
+            if (false == curl_return_code.has_value()) {
+                return ErrorCode_Failure;
+            }
+            if (CURLE_HTTP_RETURNED_ERROR == curl_return_code.value()) {
+                // Download has failed due to http error. This can be caused by an out-of-bound
+                // offset specified in the http header.
+                return ErrorCode_Failure;
+            }
+        }
+
+        if (false == at_least_one_byte_downloaded()) {
+            // We don't know if there are any bytes coming in yet
+            return ErrorCode_NotReady;
+        }
+    }
+
+    pos = m_file_pos;
+    return ErrorCode_Success;
+}
+
+auto NetworkReader::buffer_downloaded_data(NetworkReader::BufferView data) -> size_t {
+    if (data.empty()) {
+        return 0;
+    }
+    if (false == at_least_one_byte_downloaded()) {
+        m_at_least_one_byte_downloaded.store(true);
+    }
+    auto const num_bytes_to_write{data.size()};
+    while (false == data.empty()) {
+        acquire_empty_buffer();
+        if (false == m_curr_downloader_buf.has_value()) {
+            return 0;
+        }
+        auto& curr_downloader_buf{m_curr_downloader_buf.value()};
+
+        // Copy enough to fill the downloader's buf or to exhaust the data
+        auto const data_view{data.subspan(0, std::min(curr_downloader_buf.size(), data.size()))};
+        std::copy(data_view.begin(), data_view.end(), curr_downloader_buf.begin());
+        data = data.subspan(data_view.size());
+        curr_downloader_buf = curr_downloader_buf.subspan(data_view.size());
+        if (curr_downloader_buf.empty()) {
+            enqueue_filled_buffer();
+        }
+    }
+    return num_bytes_to_write;
+}
+
+auto NetworkReader::DownloaderThread::thread_method() -> void {
+    try {
+        CurlDownloadHandler curl_handler{
+                curl_progress_callback,
+                curl_write_callback,
+                static_cast<void*>(&m_reader),
+                m_reader.m_src_url,
+                m_offset,
+                m_disable_caching,
+                m_reader.m_connection_timeout,
+                m_reader.m_overall_timeout
+        };
+        auto const retcode{curl_handler.perform()};
+        m_reader.m_atomic_curl_return_code.store(retcode);
```

Comment:
Since we depend on the behaviour that when `m_curl_ret_code` is set, then `m_state` must be set, how about this:

* Let's create a method like `auto set_download_completion_status(CURLcode curl_code) -> void;` that sets the state based on the code.
  * Note that in the method, `m_curl_ret_code` must be set before `m_state`, and we should have a comment indicating that.
* Let's remove `set_state_code`.
* Let's add a comment above `m_state` and `m_curl_ret_code`: `// These should only be set from set_download_completion_status`.